promxy-operator/internal/controller/promxyservergroup_controller.go

```diff
@@ -0,0 +1,180 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+
+	coreV1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	kofv1alpha1 "github.com/k0rdent/kof/promxy-operator/api/v1alpha1"
+)
+
+const SecretNameLabel = "k0rdent.mirantis.com/promxy-secret-name"
+
+type PromxyConfigReloadFunc func() error
+
+// PromxyServerGroupReconciler reconciles a PromxyServerGroup object
+type PromxyServerGroupReconciler struct {
+	client.Client
+	Scheme             *runtime.Scheme
+	RemoteWriteUrl     string
+	PromxyConfigReload PromxyConfigReloadFunc
+}
+
+// +kubebuilder:rbac:groups=kof.k0rdent.mirantis.com,resources=promxyservergroups,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=kof.k0rdent.mirantis.com,resources=promxyservergroups/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=kof.k0rdent.mirantis.com,resources=promxyservergroups/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the PromxyServerGroup object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.19.0/pkg/reconcile
+func (r *PromxyServerGroupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	promxyServerGroupsList := &kofv1alpha1.PromxyServerGroupList{}
+	opts := []client.ListOption{
+		client.InNamespace(req.Namespace),
+	}
+
+	if err := r.List(ctx, promxyServerGroupsList, opts...); err != nil {
+		log.Error(err, "cannot get promxy server group list")
+		return ctrl.Result{}, err
+	}
+
+	promxyServerGroupsBySecretName := make(map[string][]*kofv1alpha1.PromxyServerGroup)
+
+	for _, promxyServerGroup := range promxyServerGroupsList.Items {
+		name, ok := promxyServerGroup.Labels[SecretNameLabel]
+		if !ok {
+			log.Info("Skipping promxyServerGroup that doesn't have secret name label", "promxyServerGroup", promxyServerGroup)
+			continue
+		}
+		groups, ok := promxyServerGroupsBySecretName[name]
+		if !ok {
+			groups = make([]*kofv1alpha1.PromxyServerGroup, 0)
+			groups = append(groups, &promxyServerGroup)
+		}
```

Comment:
Current code leads to only one group per secret name.

Fix:
```suggestion
		}
		groups = append(groups, &promxyServerGroup)
```

Reply:
Thanks, fixed